pkg/cmd/util/clientcmd/factory.go

```diff
@@ -774,3 +775,51 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
 	c.clients[config.GroupVersion.String()] = client
 	return client, nil
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(f *Factory) ([]unversioned.GroupResource, error) {
+	m, _ := f.Object(false)
+	oclient, _, err := f.Clients()
+	if err != nil {
+		return nil, err
+	}
+	set := make(map[string]unversioned.GroupResource)
+	all, err := oclient.Discovery().ServerResources()
+	if err != nil {
+		return nil, err
+	}
+	for apiVersion, v := range all {
+		gv, err := unversioned.ParseGroupVersion(apiVersion)
+		if err != nil {
+			continue
+		}
+		for _, r := range v.APIResources {
+			// ignore subresources
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			// because discovery info doesn't tell us whether the object is virtual or not, perform a lookup
+			// by the kind for resource (which should be the canonical resource) and then verify that the reverse
+			// lookup (KindsFor) does not error.
+			if mapping, err := m.RESTMapping(unversioned.GroupKind{Group: gv.Group, Kind: r.Kind}, gv.Version); err == nil {
+				if _, err := m.KindsFor(mapping.GroupVersionKind.GroupVersion().WithResource(mapping.Resource)); err == nil {
+					set[mapping.Resource] = unversioned.GroupResource{Group: gv.Group, Resource: mapping.Resource}
+				}
+			}
+		}
+	}
+	keys := sets.NewString()
+	for k := range set {
```

Comment:
> Not sure what you mean

`keys = sets.StringKeySet(set)`, though as I comment above, I think you can just restructure that map.